src/group_theory/semidirect_product.lean

```diff
@@ -0,0 +1,134 @@
+/-
+Copyright (c) 2020 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+-/
+import data.equiv.mul_add
+/-!
+# Semidirect product
+
+This file defines semidirect products of groups, and the canonical maps in and out of the
+semidirect product. The semidirect product of `N` and `G` given a hom `φ` from
+`φ` from `G` to the automorphism group of `N` is the product of sets with the group
+`⟨n₁, g₁⟩ * ⟨n₂, g₂⟩ = ⟨n₁ * φ g₁ n₂, g₁ * g₂⟩`
+
+## Key definitions
+
+There are two homs into the semidirect product `inl : N →* N ⋊[φ] G` and
+`inr : G →* N ⋊[φ] G`, and `lift` can be used to define maps `N ⋊[φ] G →* H`
+out of the semidirect product given maps `f₁ : N →* H` and `f₂ : G →* H` that satisfy the
+condition `∀ n g, f₁ (φ g n) = f₂ g * f₁ n * f₂ g⁻¹`
+
+## Notation
+
+This file introduces the global notation `N ⋊[φ] G` for `semidirect_product N G φ`
+
+## Tags
+group, semidirect product
+-/
+variables (N : Type*) (G : Type*) {H : Type*} [group N] [group G] [group H]
+
+/-- The semidirect product of groups `N` and `G`, given a map `φ` from `G` to the automorphism
+  group of `N`. It the product of sets with the group operation
+  `⟨n₁, g₁⟩ * ⟨n₂, g₂⟩ = ⟨n₁ * φ g₁ n₂, g₁ * g₂⟩` -/
+@[ext, derive decidable_eq]
+structure semidirect_product (φ : G →* mul_aut N) :=
+(left : N) (right : G)
+
+attribute [pp_using_anonymous_constructor] semidirect_product
+
+notation N` ⋊[`:35 φ:35`] `:0 G :35 := semidirect_product N G φ
+
+namespace semidirect_product
+
+variables {N G} {φ : G →* mul_aut N}
+
+private def one_aux : N ⋊[φ] G := ⟨1, 1⟩
+private def mul_aux (a b : N ⋊[φ] G) : N ⋊[φ] G := ⟨a.1 * φ a.2 b.1, a.right * b.right⟩
```

Comment:
You sometimes use `a.1` instead of `a.left`, also elsewhere. I don't really mind which syntactic form we use, but I think it's good to be consistent.